File: 프로그래머스/2/42747. H－Index/H－Index.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> citations) {
    int answer = 0;

    int MIN = 0, MAX = *max_element(citations.begin(), citations.end());
    int mid;
    while (MIN <= MAX) {
        mid = (MIN + MAX) / 2;
        int count = 0;
        for (int c : citations) {
            if (c >= mid) count++;
        }
        if (count >= mid && citations.size() - count <= mid) {
            answer = mid;
            MIN = mid + 1;
        }
        else MAX = mid - 1;
    }

    return answer;
}

File: 프로그래머스/2/42747. H－Index/H－Index.swift
import Foundation

func solution(_ citations:[Int]) -> Int {
    var (l, r) = (0, citations.max()!)
    var answer: Int = 0 
    
    while l <= r {
        let m = (l + r) / 2
        
        if citations.filter { $0 >= m }.count >= m {
            answer = max(answer, m)
            l = m + 1
        } else {
            r = m - 1
        }
    }
    return answer
}